pkg/store/bucket.go

```diff
@@ -3160,6 +3160,10 @@ type bucketChunkReader struct {
 	mtx        sync.Mutex
 	stats      *queryStats
 	chunkBytes []*[]byte // Byte slice to return to the chunk pool on close.
+
+	loadingChunksMtx  sync.Mutex
+	loadingChunks     bool
+	finishLoadingChks chan struct{}
```

Comment:
```suggestion
	loadingChunksCond  *sync.Cond
	loadingChunks      bool
```

Reply:
I think sync.Cond is hard to use and understand, and I would prefer to have a simpler solution with a channel.

Reply:
@fpetkovski that's also a great idea. If we bump our minimum Go version high enough, we could use generic version of atomics to easily swap something of any type.

Reply:
But I guess the critical part here is "waiting for a condition to be fulfilled": lock released when chunks are finally loaded.

Reply:
An atomic variable won't work because we must wait until the chunk loading finishes. `sync.Cond` doesn't work because at the end of `load()` it will signal only once whereas we need a permanent state of either on or off. As far as I can tell, with your suggestion in a normal operation `Close()` will hang forever because it will never receive a signal. 

I think the only way to simplify this is to get rid of the bool and create the channel under the lock. :thinking:

Reply:
In a way, the `sync.Cond` is only abstracting the channel. We need still the condition's bool variable to be checked.

Reply:
Mhm, so maybe we can merge this now and clean up later to unblock the release?

Reply:
Let's go ahead with this then, and iterate on it. 🙂